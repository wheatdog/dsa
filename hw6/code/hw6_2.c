#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "avl.h"
#include "bst.h"
#include "rb.h"

struct word 
{
    char s[130];
    int length;
};

void ReadWord(struct word *element)
{
    fgets(element->s, 129, stdin);
    int found = 0;
    for (int count = 0; count < 129 && !found; count++)
    {
        if (element->s[count] == '\n')
        {
            element->s[count] = '\0';
            element->length = count;
            found = 1;
        }
    }
}

void preorder_avl(const struct avl_node *node)
{
    if (node == NULL)
        return;
    printf("%s ", ((struct word *)node->avl_data)->s);
    if (node->avl_link[0] != NULL || node->avl_link[1] != NULL)
    {
        putchar('(');
        preorder_avl(node->avl_link[0]);
        putchar(',');
        putchar(' ');
        preorder_avl(node->avl_link[1]);
        putchar(')');
    }
}

void preorder_bst(const struct bst_node *node)
{
    if (node == NULL)
        return;
    printf("%s ", ((struct word *)node->bst_data)->s);
    if (node->bst_link[0] != NULL || node->bst_link[1] != NULL)
    {
        putchar('(');
        preorder_bst(node->bst_link[0]);
        putchar(',');
        putchar(' ');
        preorder_bst(node->bst_link[1]);
        putchar(')');
    }
}

void preorder_rb(const struct rb_node *node)
{
    if (node == NULL)
        return;
    printf("%s ", ((struct word *)node->rb_data)->s);
    if (node->rb_link[0] != NULL || node->rb_link[1] != NULL)
    {
        putchar('(');
        preorder_rb(node->rb_link[0]);
        putchar(',');
        putchar(' ');
        preorder_rb(node->rb_link[1]);
        putchar(')');
    }
}

int Compare_by_lexicographical_order(const void *pa, const void *pb, void *param)
{
    const struct word *a = (const struct word *)pa;
    const struct word *b = (const struct word *)pb;

    return strcmp(a->s, b->s);
}

int main() 
{
    struct avl_table *avltree;
    struct bst_table *bsttree;
    struct rb_table *rbtree;

    avltree = avl_create(Compare_by_lexicographical_order, NULL, NULL);
    bsttree = bst_create(Compare_by_lexicographical_order, NULL, NULL);
    rbtree = rb_create(Compare_by_lexicographical_order, NULL, NULL);

    for (int i = 0; i < 32; i++)
    {
        struct word *element = 
            (struct word *)malloc(sizeof(struct word));

        ReadWord(element);

#ifdef DEBUG
        printf("Read in word: %s, length = %d\n", element->s, element->length);
#endif

        avl_probe(avltree, element);
        bst_probe(bsttree, element);
        rb_probe(rbtree, element);
    }

    preorder_avl(avltree->avl_root); puts("");
    preorder_bst(bsttree->bst_root); puts("");
    preorder_rb(rbtree->rb_root); puts("");
    return 0;
}
