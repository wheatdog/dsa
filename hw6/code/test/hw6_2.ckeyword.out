if (do (char (break (auto , case ), continue (const , default )), extern (else (double , enum ), for (float , goto ))), static (return (long (int , register ), signed (short , sizeof )), union (switch (struct , typedef ), void (unsigned , volatile (, while )))))
auto (, break (, case (, char (, const (, continue (, default (, do (, double (, else (, enum (, extern (, float (, for (, goto (, if (, int (, long (, register (, return (, short (, signed (, sizeof (, static (, struct (, switch (, typedef (, union (, unsigned (, void (, volatile (, while )))))))))))))))))))))))))))))))
do (char (break (auto , case ), continue (const , default )), if (extern (else (double , enum ), for (float , goto )), return (long (int , register ), static (signed (short , sizeof ), union (switch (struct , typedef ), void (unsigned , volatile (, while )))))))
